File: crates/ipc/src/tests/gates.rs
//! What a step says about its gates before any of them has acted.
//!
//! The rail is drawn from this, and it drew two of a live `feature` Job's four
//! steps wrong: a step gated `human_always` — where the Job halts and waits for
//! a person — rendered as a step with no check on it, because neither the
//! declaration nor the gate was on the wire at all.

use core_model::{
    AdvanceGate, CriterionId, EvidenceType, Facts, FrozenWorkflow, GamingCheck, GamingPattern, Job,
    JobId, JudgeCheck, JudgeCriterion, ManifestId, ModelName, NewJob, ResolvedStep, StepId,
    StepSeed, Title, TopLevelOrigin, Ulid, Urgency, WorkflowId,
};

use crate::tests::{at, job};
use crate::{decode, encode, DeclaredJudge, JobDetail};

/// The four steps of the `feature` workflow as they were declared on the day
/// the rail drew two of them wrong: a judged step, a judged step with a panel
/// and a second look, a step gated on a person, and a step gated on neither.
fn gated_workflow() -> FrozenWorkflow {
    let criterion = |id: &str| JudgeCriterion {
        criterion_id: CriterionId::new(id),
        question: "does the diff do what the step said it would".to_string(),
    };
    let step = |id: &str, gate: AdvanceGate, judge: Vec<JudgeCheck>| {
        ResolvedStep::frozen(
            StepId::new(id),
            id.to_string(),
            Some(EvidenceType::Diff),
            Vec::new(),
            gate,
            judge,
            None,
        )
    };
    FrozenWorkflow::frozen(
        WorkflowId::carried(Ulid::carried("01WF")),
        "feature".to_string(),
        1,
        vec![
            step("scope", AdvanceGate::Auto, Vec::new()),
            step(
                "implement",
                AdvanceGate::AutoIfJudgePasses,
                vec![JudgeCheck::declared(
                    None,
                    1,
                    vec![criterion("c1"), criterion("c2")],
                    None,
                )],
            ),
            step(
                "tests",
                AdvanceGate::AutoIfJudgePasses,
                vec![JudgeCheck::declared(
                    None,
                    3,
                    vec![criterion("c3")],
                    Some(GamingCheck::declared(
                        None,
                        vec![GamingPattern::AssertionWeakened],
                    )),
                )],
            ),
            // Declared, and inert: no criterion and no pattern is how the
            // domain spells a check that was switched off.
            step(
                "handoff",
                AdvanceGate::HumanAlways,
                vec![JudgeCheck::declared(None, 1, Vec::new(), None)],
            ),
        ],
    )
}

fn gated_job() -> Job {
    let seed = |id: &str, ordinal: u32| StepSeed {
        step_id: StepId::new(id),
        ordinal,
    };
    Job::create_top_level(
        NewJob {
            id: JobId::carried(Ulid::carried("01FEAT")),
            title: Title::new("add the thing").expect("a title"),
            workflow: gated_workflow(),
            owner_manifest_id: ManifestId::carried(Ulid::carried("01MF")),
            urgency: Urgency::Normal,
            atomic: false,
            model: ModelName::new("a-model").expect("a model name"),
            acceptance_criteria: Vec::new(),
            steps: vec![
                seed("scope", 0),
                seed("implement", 1),
                seed("tests", 2),
                seed("handoff", 3),
            ],
            dependencies: Vec::new(),
            gate_manifests: Vec::new(),
            write_targets: None,
            subject: None,
            redispatched_from: None,
            facts: Facts::new(""),
            scope_revisions: Vec::new(),
            attachments: Vec::new(),
        },
        TopLevelOrigin::Manual,
        at("2026-08-26T09:00:00.000Z"),
    )
}

/// **The defect this closed, on the step it was worst on.** `handoff` is gated
/// `human_always`: the Job halts there and waits for a person. Nothing on the
/// wire said so, so a rail drew the commonest halt in the fleet as a step with
/// no check on it — the opposite of what happens.
#[test]
fn a_step_says_it_will_stop_for_a_person_before_it_stops() {
    let detail = JobDetail::of(&gated_job(), None, None, &[], None, None);
    let json = encode(&detail).expect("a detail is plain data");

    let gates: Vec<&str> = detail
        .steps
        .iter()
        .map(|step| {
            step.advance_gate
                .expect("the workflow declares it")
                .as_wire()
        })
        .collect();
    assert_eq!(
        gates,
        vec![
            "auto",
            "auto_if_judge_passes",
            "auto_if_judge_passes",
            "human_always"
        ],
        "every step names what it takes to get past it: {json}"
    );
    assert_eq!(
        decode::<JobDetail>("a Job in full", json.as_bytes()).expect("it round-trips"),
        detail
    );
}

/// A declared judge check crosses as counts, and the panel is named only where
/// there is one.
#[test]
fn a_declared_judge_check_crosses_as_counts_and_a_panel_only_above_one() {
    let detail = JobDetail::of(&gated_job(), None, None, &[], None, None);
    let judged: Vec<&Vec<DeclaredJudge>> = detail
        .steps
        .iter()
        .map(|step| {
            step.judge_checks
                .as_ref()
                .expect("the workflow declares it")
        })
        .collect();

    assert!(judged[0].is_empty(), "`scope` asks the Judge nothing");
    assert_eq!(judged[1].len(), 1);
    assert_eq!(judged[1][0].criteria, 2);
    assert_eq!(
        judged[1][0].panel_size, None,
        "one judge is the default and is not a panel"
    );
    assert!(!judged[1][0].gaming_check);
    assert_eq!(judged[2][0].criteria, 1);
    assert_eq!(judged[2][0].panel_size, Some(3));
    assert!(
        judged[2][0].gaming_check,
        "the second look rides along and is worth saying so"
    );
    assert!(
        judged[3].is_empty(),
        "a check with no criterion and no pattern calls no Judge, so it is not one"
    );

    let json = encode(&detail).expect("a detail is plain data");
    assert!(
        !json.contains("does the diff do what the step said"),
        "a criterion's question is a prompt and does not cross: {json}"
    );
    assert!(
        !json.contains("\"panel_size\":1"),
        "absent at one, never present-and-default: {json}"
    );
}

/// **Empty and absent are two sentences here too.** A step Fleet holds no
/// workflow for carries neither key, so a client cannot read "this Fleet cannot
/// say" as "nothing will happen".
#[test]
fn a_step_the_workflow_does_not_declare_carries_neither_key() {
    let job = job();
    let declared = encode(&JobDetail::of(&job, None, None, &[], None, None)).expect("plain data");
    assert!(
        declared.contains("\"judge_checks\":[]") && declared.contains("\"advance_gate\":\"auto\""),
        "a declared step answers both: {declared}"
    );

    let mut detail = JobDetail::of(&job, None, None, &[], None, None);
    detail.steps[0].judge_checks = None;
    detail.steps[0].advance_gate = None;
    let unanswerable = encode(&detail).expect("plain data");
    assert!(
        !unanswerable.contains("judge_checks") && !unanswerable.contains("advance_gate"),
        "absent, never present-and-null: {unanswerable}"
    );
    assert_eq!(
        decode::<JobDetail>("a Job in full", unanswerable.as_bytes()).expect("it round-trips"),
        detail
    );
}

File: packages/screens/src/gates.ts
// What a step's gates found, read once for both surfaces that draw it.
//
// **Two surfaces read `check_runs` and `judged` now, and one of them is new.**
// The phase strip says where a step stands; the Checks and Verdicts chapters
// say what the gates found and let a person argue with it. Those are the same
// rows read for two purposes, and reading them twice is the defect `#321`
// already found once in this file's neighbours — two surfaces stating one
// ordering separately, agreeing until one of them changed.
//
// **Nothing here is JSX, and that is the seam.** What a strip row looks like
// and what a verdict grid looks like are each surface's own answer. What they
// must not disagree about is which attempt's rows count, which member of a
// panel answered what, and whether a criterion was refused — so that is what
// lives here and nothing else does.
//
// The three rules this file holds, each of them from somewhere else:
//
// | Rule | Where it comes from |
// |---|---|
// | Only the current attempt's rows are the live gate | `check_runs` and `judged` hold every attempt since protocol 7.0 |
// | A panel is one row per member and reads as one row per criterion | `Judged.member`, since protocol 7.7 |
// | Any single refusal refuses the criterion | `docs/concepts/judge.md`, unanimity |

import { CHECK_ADVANCES, CRITERION_VERDICT_CHECK } from "@armada/components";
import type { CheckRun, Criterion, DeclaredCheck, Judged, StepDetail } from "@armada/protocol";

import { isSweepMarker, nameOf } from "./declared";
import { onlyCurrentAttempt } from "./facts";

/**
 * One declared Check, and what this attempt's run of it came to.
 *
 * **Declared first, run second.** The list is the step's declaration, so a
 * Check the gate has not reached keeps its row — the shape of what is still
 * coming is part of reading a running Job, and a list built from the runs
 * would make a Job look like it has fewer gates than it has.
 */
export type CheckRead = {
  /** The Check's name, or the built-in's kind. What a citation resolves against. */
  name: string;
  /** The declaration, for the command and the paths it covers. */
  check: DeclaredCheck;
  /** Absent where this attempt's gate has not reached it. */
  run: CheckRun | undefined;
};

/**
 * The step's Checks, each beside what it came to on the current attempt.
 *
 * **Empty is a step that gates on nothing, and so is a step whose workflow
 * Fleet does not hold.** Those are two different sentences and the phase
 * strip's note is where they are told apart — nothing that draws a list of
 * Checks can say either of them, so this answers with no rows and lets the
 * caller draw nothing rather than an empty region.
 */
export function checksOf(step: StepDetail): CheckRead[] {
  // The sweep marker never counts as a Check and never draws as one — it
  // declares that the step gates on everything the repository declares, and
  // nothing ever runs it. `run.ts`'s tier count reads this same filter, on
  // this same list, for the reason atop this file.
  const declared = (step.checks ?? []).filter((check) => !isSweepMarker(check));
  const runs = onlyCurrentAttempt(step, step.check_runs);
  return declared.map((check) => {
    const name = nameOf(check);
    return { name, check, run: runs.find((ran) => ran.name === name) };
  });
}

/** A Check that did not pass, read off the registry's own `advances`. */
export function didNotPass(run: CheckRun): boolean {
  return CHECK_ADVANCES[run.outcome] === false;
}

/**
 * What a tier the run has not got to stands at.
 *
 * **The registry's word, and `criterion_verdict_check` owns it**: the five
 * `check_outcome` carries are what a Check that *ran* did, and `icons.toml`
 * already calls "not reached" a Check state. Once here for every surface that
 * says it, rather than typed at each.
 */
export const NOT_REACHED = CRITERION_VERDICT_CHECK.not_reached?.verb ?? "not_reached";

/** The Checks that ran on this attempt, and the ones among them that failed. */
export function howTheChecksWent(reads: readonly CheckRead[]): {
  ran: CheckRead[];
  failed: CheckRead[];
} {
  const ran = reads.filter((read) => read.run !== undefined);
  return { ran, failed: ran.filter((read) => read.run !== undefined && didNotPass(read.run)) };
}

/**
 * What the Checks tier stands at, in words — `3 of 3 passed`.
 *
 * **One sentence, because two surfaces say it.** The phase strip's tier and the
 * Checks chapter's header sit four inches apart, and a count written twice is a
 * count that disagrees with itself the day `skipped` stops advancing a step.
 */
export function checksStand(reads: readonly CheckRead[]): string {
  const { ran, failed } = howTheChecksWent(reads);
  if (ran.length === 0) return NOT_REACHED;
  return failed.length > 0
    ? `${failed.length} of ${reads.length} did not pass`
    : `${ran.length} of ${reads.length} passed`;
}

/**
 * Which Check's output an `o` press opens on this step, or none.
 *
 * **The failed one first.** A person reaching for an output on a step that
 * stopped wants the Check that says why; on a step where nothing failed there
 * is still a reading worth opening, so the first output there is answers
 * instead of nothing.
 *
 * **Narrowed to the current attempt, like everything else here.** Read across
 * every attempt this surfaced a stale run's output the moment a step had been
 * worked more than once — and the key and the Checks chapter's own act both
 * come through here, so the two cannot open different files.
 */
export function outputOf(step: StepDetail | undefined): string | undefined {
  return outputRunOf(step)?.output_path;
}

/**
 * Which Check's output a reader is offered on this step, whole.
 *
 * **One reading behind three surfaces.** The header act opens the file, `o`
 * opens the file, and the chapter now reads it into the panel — three ways of
 * asking one question, and three answers to it would put a person reading one
 * Check's output under a button that opens another's. `outputOf` is this,
 * narrowed to the path, and exists because two callers only ever wanted that.
 *
 * The row rather than the path, because a reader has to say whose output it is
 * showing: a console pane with a file name and no Check name is a transcript
 * nobody can attribute.
 */
export function outputRunOf(step: StepDetail | undefined): CheckRun | undefined {
  if (step === undefined) return undefined;
  const runs = onlyCurrentAttempt(step, step.check_runs).filter(
    (run) => run.output_path !== undefined,
  );
  return runs.find(didNotPass) ?? runs[0];
}

/**
 * One criterion, and what the panel that answered it made of it.
 *
 * **A panel sends one row per member and a person reads one row per
 * criterion.** At `panel_size: 3` three rows arrive carrying the same
 * `criterion_id` and differing only in `member`; drawn straight they are the
 * same sentence three times.
 */
export type Panel = {
  criterionId: string;
  /**
   * The Job's own criterion, where the id joins. **Absent is a criterion this
   * Job does not carry** — a verdict on something the frozen list has no row
   * for, which is a fact to draw rather than one to hide.
   */
  criterion: Criterion | undefined;
  /**
   * The criterion's frozen 1-based position in `acceptance_criteria`. What a
   * citation to `02` means, and never the row's place on screen. Absent where
   * the id does not join, because a position cannot be invented for it.
   */
  ordinal: number | undefined;
  /**
   * Every member that answered, in panel order. **One entry at `panel_size:
   * 1`**, whose `member` is absent — the convention `Judged.member` keeps.
   */
  members: Judged[];
  /** The members that refused. Empty is a criterion nothing objected to. */
  refused: Judged[];
  /** Unanimity, from `docs/concepts/judge.md`: one veto refuses the criterion. */
  verdict: "met" | "not_met";
};

/**
 * A step's verdicts as one entry per criterion, in the order they were asked.
 *
 * **First-appearance order, never sorted.** The order is the order the criteria
 * were asked, which is the order they were written, and a criterion may be
 * appended but never reordered — so position is stable and is what a citation
 * to `02` means.
 *
 * **Narrowed to the current attempt.** `judged` holds every attempt's rows
 * since protocol 7.0, and both surfaces draw the live gate.
 */
export function panelsOf(step: StepDetail, criteria: readonly Criterion[]): Panel[] {
  const held = new Map<string, Judged[]>();
  for (const one of onlyCurrentAttempt(step, step.judged)) {
    const already = held.get(one.criterion_id);
    if (already === undefined) held.set(one.criterion_id, [one]);
    else already.push(one);
  }
  return [...held].map(([criterionId, answered]) => {
    // In panel order rather than in arrival order. `member` is a position, and
    // a grid whose columns changed order between two criteria would put one
    // judge's mark under another judge's heading.
    const members = [...answered].sort((a, b) => (a.member ?? 1) - (b.member ?? 1));
    const at = criteria.findIndex((held) => held.criterion_id === criterionId);
    const refused = members.filter((one) => one.verdict !== "met");
    return {
      criterionId,
      criterion: at === -1 ? undefined : criteria[at],
      ordinal: at === -1 ? undefined : at + 1,
      members,
      refused,
      verdict: refused.length === 0 ? "met" : "not_met",
    };
  });
}

/** How many criteria the step's declaration says the panel will answer. */
export function askedOf(step: StepDetail): number {
  return (step.judge_checks ?? []).reduce((sum, judge) => sum + judge.criteria, 0);
}

/**
 * Whether this attempt's panel was asked and never answered, rather than
 * never asked at all. Both draw `judged` empty for the current attempt, and
 * they are not the same sentence: one is a call still ahead of the step, the
 * other is one Fleet already made and could not read back.
 *
 * **`last_verdict.trigger` and not a live `judging` read**, because a step
 * open after the Job stopped carries no in-flight call — `judging` is
 * "right now", and there is no "now" left to name once the Job is over.
 */
export function stoppedUndecided(step: StepDetail): boolean {
  return step.last_verdict?.trigger === "gate_undecided";
}

/**
 * Fleet's own sentence, corrected to this screen's case. Fleet writes what it
 * logged, lower-case and unpunctuated like every other log line; everywhere
 * this crosses onto a sentence of its own the two surfaces that draw it would
 * otherwise punctuate it two different ways.
 */
export function sentenceOf(said: string): string {
  const capped = said.charAt(0).toUpperCase() + said.slice(1);
  return /[.!?]$/.test(capped) ? capped : `${capped}.`;
}

/**
 * How many judges answer each criterion.
 *
 * **The declaration first, the rows second.** `panel_size` is what the header
 * of a verdict grid is drawn from before a single row arrives, and it is
 * **absent at one** — the convention `Judged.member` keeps, so a value always
 * means a panel. The rows are the floor under it: a panel that answered with
 * more members than were declared is a fact, not a reason to draw fewer marks
 * than there are.
 */
export function panelSizeOf(step: StepDetail, panels: readonly Panel[]): number {
  const declared = (step.judge_checks ?? []).map((judge) => judge.panel_size ?? 1);
  return Math.max(1, ...declared, ...panels.map((panel) => panel.members.length));
}

/**
 * What one member of a panel is called — `j2`.
 *
 * **A position, not a person, and never part of a citation.** `Judged.member`
 * says so, and it is absent at `panel_size: 1`, where the panel is one judge
 * and the number would be a count nobody asked for.
 */
export function judgeNamed(one: Judged): string {
  return `j${one.member ?? 1}`;
}

/**
 * A run of judges, in a sentence — `j2`, `j1 and j3`, `j1, j2 and j3`.
 *
 * Written here rather than at each caller because the Checks row, the split and
 * the silence line all name the same set and would each have found their own
 * comma.
 */
export function judgesSaid(members: readonly Judged[]): string {
  const named = members.map(judgeNamed);
  if (named.length <= 1) return named.join("");
  return `${named.slice(0, -1).join(", ")} and ${named[named.length - 1]}`;
}
